src/language_server/custom_requests/Search.ml

```diff
@@ -0,0 +1,168 @@
+(* Austin Theriault
+ *
+ * Copyright (C) 2019-2023 Semgrep, Inc.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation, with the
+ * special exception on linking described in file LICENSE.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+(* Commentary *)
+(* Handles searching workspace by Semgrep pattern *)
+
+(*****************************************************************************)
+(* Prelude *)
+(*****************************************************************************)
+
+open Yojson.Safe.Util
+open Jsonrpc
+open Common
+open Lsp
+open Types
+module In = Input_to_core_t
+module Out = Output_from_core_t
+module XP = Xpattern
+module Conv = Convert_utils
+
+let logger = Logging.get_logger [ __MODULE__ ]
+(*****************************************************************************)
+(* Code *)
+(*****************************************************************************)
+
+let meth = "semgrep/search"
+
+(* Taken from osemgrep/networking/Rule_fetching.ml *)
+(* Pretty soon we should probably move LSP to OSemgrep *)
+let rules_of_pattern pat lang_opt =
+  let fk = Tok.unsafe_fake_tok "" in
+  let rule_for_xlang xlang =
+    let xpat = Parse_rule.parse_xpattern xlang (pat, fk) in
+    (* force the parsing of the pattern to get the parse error if any *)
+    (match xpat.XP.pat with
+    | XP.Sem (lpat, _) -> Lazy.force lpat |> ignore
+    | XP.Spacegrep _
+    | XP.Aliengrep _
+    | XP.Regexp _ ->
+        ());
+    let rule = Rule.rule_of_xpattern xlang xpat in
+    { rule with id = (Rule.ID.of_string "-", fk) }
+  in
+  match lang_opt with
+  | Some lang ->
+      (* TODO? capture also parse errors here? and transform the pattern
+       * parse error in invalid_rule_error to return in rules_and_origin? *)
+      [ rule_for_xlang (Xlang.of_lang lang) ]
+  | None ->
+      (* We need uniq_by because Lang.assoc contain multiple times the
```

Comment:
i see you've copied this from `Rule_fetching.ml` -- any chance that we could extract this logic to a single helper, or if that's too far beyond the scope of this PR, leave a comment coupling the two?

Reply:
There's a comment on line 40 re: this :)

Reply:
so you did! completely missed that